src/gridfs/download.rs

```diff
@@ -139,16 +142,214 @@ impl GridFsBucket {
+        Ok(Self {
+            state: initial_state,
+            cached_bytes: None,
+            current_n: 0,
+            file,
+        })
+    }
+
+    /// Gets the file `id` for the stream.
+    pub fn files_id(&self) -> &Bson {
+        &self.file.id
+    }
+}
+
+impl AsyncRead for GridFsDownloadStream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<std::result::Result<usize, futures_util::io::Error>> {
+        let stream = self.get_mut();
+
+        if let Some(cache) = &mut stream.cached_bytes {
+            let bytes_to_write = std::cmp::min(cache.len(), buf.len());
+            buf[..bytes_to_write].copy_from_slice(cache.drain(0..bytes_to_write).as_slice());
+            if cache.is_empty() {
+                stream.cached_bytes = None;
+            }
+            return Poll::Ready(Ok(bytes_to_write));
+        }
+
+        let future = match &mut stream.state {
+            State::Idle(cursor) => {
+                let chunks_in_buf =
+                    FilesCollectionDocument::n_from_vals(buf.len() as u64, stream.file.chunk_size);
+                // We should read from current_n to chunks_in_buf + current_n, or, if that would
+                // exceed the total number of chunks in the file, to the last chunk in the file.
+                let final_n = std::cmp::min(chunks_in_buf + stream.current_n, stream.file.n());
+                let n_range = stream.current_n..final_n;
+                let total_chunks = n_range.end - n_range.start;
+
+                let new_future = get_bytes(
+                    cursor.take().unwrap(),
+                    n_range,
+                    stream.file.chunk_size,
+                    stream.file.length,
+                )
+                .boxed();
+
+                stream.current_n += total_chunks;
+                stream.state.set_busy(new_future)
+            }
+            State::Busy(future) => future,
+            State::Done => return Poll::Ready(Ok(0)),
+        };
+
+        let result = match future.poll_unpin(cx) {
+            Poll::Ready(result) => result,
+            Poll::Pending => return Poll::Pending,
+        };
+
+        match result {
+            Ok((mut bytes, cursor)) => {
+                let bytes_to_write = std::cmp::min(bytes.len(), buf.len());
+                buf[..bytes_to_write].copy_from_slice(&bytes.drain(0..bytes_to_write).as_slice());
+
+                // The cache is always drained before creating a future, so we don't need to worry
+                // about overwriting anything.
+                if !bytes.is_empty() {
+                    stream.cached_bytes = Some(bytes);
+                }
+
+                stream.state = if cursor.has_next() {
+                    State::Idle(Box::new(Some(cursor)))
+                } else {
+                    State::Done
+                };
+
+                Poll::Ready(Ok(bytes_to_write))
+            }
+            Err(error) => {
```

Comment:
I originally allowed errors to be recoverable (i.e. not set the stream's state to `Done`), but after our discussion about losing partial chunks when the buffer size doesn't exactly divide the chunk size, I realized the same thing would happen if an error occurred in the middle of iterating the cursor. While we could cache what has been read so far, there is a question around what should be done with the data in malformed chunks -- should it be included in the cache, or should it be ignored? Additionally, all data in chunks that couldn't deserialize properly would be lost. Because of this I think it makes more sense to allow these errors to invalidate the stream. Our retryability logic will retry recoverable errors, so it's likely that the errors returned here will be unrecoverable anyway.